core/state_transition.go

```diff
@@ -366,9 +374,93 @@ func (st *StateTransition) applyEditValidatorTx(ev *staking.EditValidator, block
 }
 
 func (st *StateTransition) applyDelegateTx(delegate *staking.Delegate) error {
+	if !st.state.IsValidator(delegate.ValidatorAddress) {
+		return errValidatorNotExist
+	}
+	wrapper := st.state.GetStakingInfo(delegate.ValidatorAddress)
+	if wrapper == nil {
+		return errValidatorNotExist
+	}
+
+	stateDB := st.state
+	delegatorExist := false
+	for i := range wrapper.Delegations {
+		delegation := wrapper.Delegations[i]
+		if bytes.Equal(delegation.DelegatorAddress.Bytes(), delegate.DelegatorAddress.Bytes()) {
+			delegatorExist = true
+			totalInUndelegation := delegation.TotalInUndelegation()
+			// If the sum of normal balance and the total amount of tokens in undelegation is greater than the amount to delegate
+			if big.NewInt(0).Add(totalInUndelegation, stateDB.GetBalance(delegate.DelegatorAddress)).Cmp(delegate.Amount) >= 0 {
+				// Firstly use the tokens in undelegation to delegate (redelegate)
+				undelegateAmount := big.NewInt(0).Set(delegate.Amount)
+				// Use the latest undelegated token first as it has the longest remaining locking time.
+				i := len(delegation.Entries) - 1
```

Comment:
How we know this invariant (latest has longest remaining locking time) is always true? Can sorting list in some way make us keep the invariant?

Reply:
Because we are using this list always as a FIFO queue. It's always ordered by time.

Reply:
But added a TODO to sort the list just in case.